apiserver/internal/handlers/objects/charms.go

```diff
@@ -0,0 +1,400 @@
+	if err != nil {
+		return errors.Capture(err)
+	}
+
+	// Add a charm to the store provider.
+	charmURL, err := h.processPut(ctx, r, st, applicationService)
+	if err != nil {
+		return jujuerrors.NewBadRequest(err, "")
+	}
+	headers := map[string]string{
+		params.JujuCharmURLHeader: charmURL.String(),
+	}
+	return errors.Capture(sendStatusAndHeadersAndJSON(w, http.StatusOK, headers, &params.CharmsResponse{CharmURL: charmURL.String()}))
+}
+
+func (h *ObjectsCharmHTTPHandler) processPut(ctx context.Context, r *http.Request, st ModelState, applicationService ApplicationService) (*charm.URL, error) {
+	name, shaFromQuery, err := splitNameAndSHAFromQuery(r.URL.Query())
+	if err != nil {
+		return nil, errors.Capture(err)
+	}
+
+	curlStr := r.Header.Get(params.JujuCharmURLHeader)
+	if curlStr == "" {
+		return nil, jujuerrors.BadRequestf("missing %q header", params.JujuCharmURLHeader)
+	}
+	curl, err := charm.ParseURL(curlStr)
+	if err != nil {
+		return nil, jujuerrors.BadRequestf("%q is not a valid charm url", curlStr)
+	}
+	curl.Name = name
+
+	// charmhub charms may only be uploaded into models which are being
+	// imported during model migrations. There's currently no other time
+	// where it makes sense to accept repository charms through this
+	// endpoint.
+	// TODO (stickupkid): This should be moved to the application service once
+	// model migration is complete.
+	isImporting, err := modelIsImporting(st)
+	if err != nil {
+		return nil, errors.Capture(err)
+	}
+
+	var source corecharm.Source
+	switch curl.Schema {
+	case charm.CharmHub.String():
+		source = corecharm.CharmHub
+	case charm.Local.String():
+		source = corecharm.Local
+	default:
+		return nil, jujuerrors.BadRequestf("unsupported charm source %q", curl.Schema)
+	}
+
+	locator, err := applicationService.ResolveUploadCharm(ctx, applicationcharm.ResolveUploadCharm{
+		Name:         curl.Name,
+		Revision:     curl.Revision,
+		Source:       source,
+		Architecture: curl.Architecture,
+		SHA256Prefix: shaFromQuery,
+
+		// Prevent an upload starvation attack by limiting the size of the
+		// charm that can be uploaded.
+		Reader: io.LimitReader(r.Body, maxUploadSize),
+
+		// Importing indicates that the charm is being uploaded during model
+		// migration import. This is useful to set the provenance of the charm
+		// correctly.
+		Importing: isImporting,
+	})
+	if errors.Is(err, applicationerrors.CharmNotFound) {
+		logger.Criticalf(context.TODO(), "charm not found on object store")
+		return nil, jujuerrors.NotFoundf("charm")
+	} else if errors.Is(err, applicationerrors.CharmAlreadyAvailable) {
+		return nil, jujuerrors.AlreadyExistsf("charm")
+	} else if err != nil {
+		return nil, errors.Capture(err)
+	}
+
+	return h.makeCharmURL(locator, curl.Architecture != "")
+}
+
+// CharmURLFromLocator returns a charm URL from a charm locator.
+// This will always include the architecture.
+func CharmURLFromLocator(locator applicationcharm.CharmLocator, _ bool) (*charm.URL, error) {
```

Comment:
```suggestion
func CharmURLFromLocator(locator applicationcharm.CharmLocator) (*charm.URL, error) {
```
We can drop the legacy "include architecture" bool, right?

Reply:
I'll check.

Reply:
No we can't. During a migration, we can't include the architecture if it's a local charm. This ensures backwards compatibility with older controllers.